File: app/src/main/java/com/toandev/gpsimage/SettingsActivity.java
package com.toandev.gpsimage;

import android.content.Intent;
import android.graphics.Bitmap;
import android.os.AsyncTask;
import android.os.Build.VERSION;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v4.content.ContextCompat;
import android.support.v7.app.ActionBar;
import android.support.v7.app.AppCompatActivity;
import android.text.InputFilter;
import android.text.InputFilter.AllCaps;
import android.util.DisplayMetrics;
import android.util.Log;
import android.view.KeyEvent;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.view.View.OnClickListener;
import android.view.ViewTreeObserver.OnGlobalLayoutListener;
import android.view.inputmethod.InputMethodManager;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.LinearLayout.LayoutParams;
import android.widget.SeekBar;
import android.widget.SeekBar.OnSeekBarChangeListener;
import android.widget.TextView;
import android.widget.TextView.OnEditorActionListener;
import android.widget.Toast;

import com.toandev.demogpsimage.R;
import com.toandev.gpsimage.utils.MyLocation;
import com.toandev.gpsimage.utils.MySharedPreferences;
import com.toandev.gpsimage.utils.MyUtils;

import net.rdrei.android.dirchooser.DirectoryChooserConfig;
import net.rdrei.android.dirchooser.DirectoryChooserFragment;

import java.io.IOException;

public class SettingsActivity extends AppCompatActivity implements OnClickListener, OnSeekBarChangeListener, DirectoryChooserFragment.OnFragmentInteractionListener {
    private EditText etInputFolder;
    private EditText etOutputFolder;
    private EditText etUnitInfo;
    private ImageView ivMap;
    private DirectoryChooserFragment mDialog;
    private MyLocation mLocation;
    private SeekBar sbImageQuality;
    private SeekBar sbMapWidth;
    private SeekBar sbMapZoom;

    private class MyDownloader extends AsyncTask<String, Bitmap, Bitmap> {
        private int mHeight;
        private double mLatitude;
        private double mLongitude;
        private int mRotate;
        private int mWidth;
        private int mZoom;

        public MyDownloader(double lat, double lon, int zoom, int mapWidth, int mapHeight, int rotate) {
            this.mLongitude = lon;
            this.mLatitude = lat;
            this.mZoom = zoom;
            this.mWidth = mapWidth;
            this.mHeight = mapHeight;
            this.mRotate = rotate;
        }

        protected Bitmap doInBackground(String... params) {
            try {
                return MyUtils.getMap(this.mLatitude, this.mLongitude, this.mZoom, this.mWidth, this.mHeight, (float) this.mRotate, SettingsActivity.this.getResources().getString(R.string.key));
            } catch (IOException e) {
                return null;
            }
        }

        protected void onPostExecute(Bitmap bitmap) {
            super.onPostExecute(bitmap);
            Log.i("hson", "onpost execute");
            if (bitmap == null) {
                Toast.makeText(SettingsActivity.this, SettingsActivity.this.getString(R.string.alert_cannot_connect_internet), Toast.LENGTH_LONG).show();
                Log.i("hson", "show toast");
                return;
            }
            SettingsActivity.this.ivMap.setImageBitmap(bitmap);
            Log.i("hson", "show image");
        }
    }

    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_settings);
        ActionBar actionbar = getSupportActionBar();
        if (actionbar != null) {
            actionbar.setDisplayHomeAsUpEnabled(true);
        }
        this.mLocation = new MyLocation(this);
        this.etInputFolder = (EditText) findViewById(R.id.etInputFolder);
        this.etOutputFolder = (EditText) findViewById(R.id.etOutputFolder);
        this.etUnitInfo = (EditText) findViewById(R.id.etUnitName);
        this.sbMapWidth = (SeekBar) findViewById(R.id.sbMapWidth);
        this.sbMapZoom = (SeekBar) findViewById(R.id.sbMapZoom);
        this.sbImageQuality = (SeekBar) findViewById(R.id.sbImageQuality);
        this.ivMap = (ImageView) findViewById(R.id.ivMap);
        MySharedPreferences pref = MySharedPreferences.Companion.getInstance(this);
        this.etInputFolder.setText(pref.getObserverFilePath());
        this.etOutputFolder.setText(pref.getOutputFilePath());
        this.etUnitInfo.setText(pref.getUnitInfo());
        this.sbMapWidth.setProgress(pref.getMapWidth() - 100);
        this.sbMapZoom.setProgress(pref.getMapZoom());
        this.sbImageQuality.setProgress(pref.getImageQuality());
        this.etInputFolder.setOnClickListener(this);
        this.etOutputFolder.setOnClickListener(this);
        this.etUnitInfo.setFilters(new InputFilter[]{new AllCaps()});
        this.etUnitInfo.setOnEditorActionListener(new OnEditorActionListener() {
            public boolean onEditorAction(TextView v, int actionId, KeyEvent event) {
                if (actionId == 6) {
                    SettingsActivity.this.etUnitInfo.clearFocus();
                    ((InputMethodManager) SettingsActivity.this.getSystemService("input_method")).hideSoftInputFromWindow(SettingsActivity.this.etUnitInfo.getWindowToken(), 0);
                }
                return false;
            }
        });
        this.sbMapWidth.setOnSeekBarChangeListener(this);
        this.sbMapZoom.setOnSeekBarChangeListener(this);
        DisplayMetrics metrics = new DisplayMetrics();
        getWindowManager().getDefaultDisplay().getMetrics(metrics);
        LayoutParams params = new LayoutParams(-1, metrics.widthPixels);
        params.setMargins(10, 10, 10, 10);
        this.ivMap.setLayoutParams(params);
        this.ivMap.getViewTreeObserver().addOnGlobalLayoutListener(new OnGlobalLayoutListener() {
            public void onGlobalLayout() {
                if (VERSION.SDK_INT < 16) {
                    SettingsActivity.this.ivMap.getViewTreeObserver().removeGlobalOnLayoutListener(this);
                } else {
                    SettingsActivity.this.ivMap.getViewTreeObserver().removeOnGlobalLayoutListener(this);
                }
                SettingsActivity.this.showMap();
            }
        });
    }

    protected void onResume() {
        super.onResume();
        this.mLocation.requestGPSUpdate();
    }

    protected void onStop() {
        super.onStop();
        this.mLocation.removeUpdates();
    }

    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.settings, menu);
        return true;
    }

    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.save :
                MySharedPreferences pref = MySharedPreferences.Companion.getInstance(this);
                pref.setObserverFilePath(this.etInputFolder.getText().toString());
                pref.setOutputFilePath(this.etOutputFolder.getText().toString());
                pref.setUnitInfo(this.etUnitInfo.getText().toString());
                pref.setMapWidth(this.sbMapWidth.getProgress() + 100);
                pref.setMapZoom(this.sbMapZoom.getProgress());
                pref.setImageQuality(this.sbImageQuality.getProgress());
                Intent returnIntent = new Intent();
                returnIntent.putExtra(MainActivity.Companion.getIS_UPDATE_SERVICE(), true);
                setResult(-1, returnIntent);
                Toast.makeText(this, R.string.alert_save, Toast.LENGTH_LONG).show();
                finish();
                return true;
            default:
                return super.onOptionsItemSelected(item);
        }
    }

    protected void onDestroy() {
        super.onDestroy();
        if (ContextCompat.checkSelfPermission(this, "android.permission.ACCESS_FINE_LOCATION") == 0 || ContextCompat.checkSelfPermission(this, "android.permission.ACCESS_COARSE_LOCATION") == 0) {
            this.mLocation.removeUpdates();
        }
    }

    private void showMap() {
        new MyDownloader(this.mLocation.getLatitude(), this.mLocation.getLongitude(), this.sbMapZoom.getProgress(), this.sbMapWidth.getProgress() + 100, this.ivMap.getHeight() / 2, this.sbImageQuality.getProgress()).execute(new String[0]);
    }

    public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
    }

    public void onStartTrackingTouch(SeekBar seekBar) {
    }

    public void onStopTrackingTouch(SeekBar seekBar) {
        showMap();
    }

    public void onSelectDirectory(@NonNull String path) {
        this.mDialog.dismiss();
        switch (Integer.parseInt(this.mDialog.getTag())) {
            case R.id.etInputFolder /*2131492954*/:
                this.etInputFolder.setText(path);
                return;
            case R.id.etOutputFolder /*2131492956*/:
                this.etOutputFolder.setText(path);
                return;
            default:
                return;
        }
    }

    public void onCancelChooser() {
        this.mDialog.dismiss();
    }

    public void onClick(View v) {
        switch (v.getId()) {
            case R.id.etInputFolder /*2131492954*/:
                showDialog(this.etInputFolder.getText().toString(), String.valueOf(R.id.etInputFolder));
                return;
            case R.id.etOutputFolder /*2131492956*/:
                showDialog(this.etOutputFolder.getText().toString(), String.valueOf(R.id.etOutputFolder));
                return;
            default:
                return;
        }
    }

    public void showDialog(String folderPath, String tag) {
        this.mDialog = DirectoryChooserFragment.newInstance(DirectoryChooserConfig.builder().newDirectoryName(getString(R.string.new_directory_name)).initialDirectory(folderPath).allowNewDirectoryNameModification(true).build());
        this.mDialog.show(getFragmentManager(), tag);
    }
}

File: app/src/main/java/com/toandev/gpsimage/SettingsActivity.kt
package com.toandev.gpsimage

import android.content.Intent
import android.graphics.Bitmap
import android.os.AsyncTask
import android.os.Build.VERSION
import android.os.Bundle
import android.support.v4.content.ContextCompat
import android.support.v7.app.ActionBar
import android.support.v7.app.AppCompatActivity
import android.text.InputFilter
import android.text.InputFilter.AllCaps
import android.util.DisplayMetrics
import android.util.Log
import android.view.KeyEvent
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.view.View.OnClickListener
import android.view.ViewTreeObserver.OnGlobalLayoutListener
import android.view.inputmethod.InputMethodManager
import android.widget.EditText
import android.widget.ImageView
import android.widget.LinearLayout.LayoutParams
import android.widget.SeekBar
import android.widget.SeekBar.OnSeekBarChangeListener
import android.widget.TextView
import android.widget.TextView.OnEditorActionListener
import android.widget.Toast

import com.toandev.demogpsimage.R
import com.toandev.gpsimage.utils.MyLocation
import com.toandev.gpsimage.utils.MySharedPreferences
import com.toandev.gpsimage.utils.MyUtils

import net.rdrei.android.dirchooser.DirectoryChooserConfig
import net.rdrei.android.dirchooser.DirectoryChooserFragment

import java.io.IOException

class SettingsActivity : AppCompatActivity(), OnClickListener, OnSeekBarChangeListener, DirectoryChooserFragment.OnFragmentInteractionListener {
    private var etInputFolder: EditText? = null
    private var etOutputFolder: EditText? = null
    private var etUnitInfo: EditText? = null
    private var ivMap: ImageView? = null
    private var mDialog: DirectoryChooserFragment? = null
    private var mLocation: MyLocation? = null
    private var sbImageQuality: SeekBar? = null
    private var sbMapWidth: SeekBar? = null
    private var sbMapZoom: SeekBar? = null

    private inner class MyDownloader(private val mLatitude: Double, private val mLongitude: Double, private val mZoom: Int, private val mWidth: Int, private val mHeight: Int, private val mRotate: Int) : AsyncTask<String, Bitmap, Bitmap>() {

        override fun doInBackground(vararg params: String): Bitmap? {
            try {
                return MyUtils.getMap(this.mLatitude, this.mLongitude, this.mZoom, this.mWidth, this.mHeight, this.mRotate.toFloat(), this@SettingsActivity.resources.getString(R.string.key))
            } catch (e: IOException) {
                return null
            }

        }

        override fun onPostExecute(bitmap: Bitmap?) {
            super.onPostExecute(bitmap)
            Log.i("hson", "onpost execute")
            if (bitmap == null) {
                Toast.makeText(this@SettingsActivity, this@SettingsActivity.getString(R.string.alert_cannot_connect_internet), Toast.LENGTH_LONG).show()
                Log.i("hson", "show toast")
                return
            }
            this@SettingsActivity.ivMap!!.setImageBitmap(bitmap)
            Log.i("hson", "show image")
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_settings)
        val actionbar = supportActionBar
        actionbar?.setDisplayHomeAsUpEnabled(true)
        this.mLocation = MyLocation(this)
        this.etInputFolder = findViewById<View>(R.id.etInputFolder) as EditText
        this.etOutputFolder = findViewById<View>(R.id.etOutputFolder) as EditText
        this.etUnitInfo = findViewById<View>(R.id.etUnitName) as EditText
        this.sbMapWidth = findViewById<View>(R.id.sbMapWidth) as SeekBar
        this.sbMapZoom = findViewById<View>(R.id.sbMapZoom) as SeekBar
        this.sbImageQuality = findViewById<View>(R.id.sbImageQuality) as SeekBar
        this.ivMap = findViewById<View>(R.id.ivMap) as ImageView
        val pref = MySharedPreferences.getInstance(this)
        this.etInputFolder!!.setText(pref.observerFilePath)
        this.etOutputFolder!!.setText(pref.outputFilePath)
        this.etUnitInfo!!.setText(pref.unitInfo)
        this.sbMapWidth!!.progress = pref.mapWidth - 100
        this.sbMapZoom!!.progress = pref.mapZoom
        this.sbImageQuality!!.progress = pref.imageQuality
        this.etInputFolder!!.setOnClickListener(this)
        this.etOutputFolder!!.setOnClickListener(this)
        this.etUnitInfo!!.filters = arrayOf<InputFilter>(AllCaps())
        this.etUnitInfo!!.setOnEditorActionListener { v, actionId, event ->
            if (actionId == 6) {
                this@SettingsActivity.etUnitInfo!!.clearFocus()
                (this@SettingsActivity.getSystemService("input_method") as InputMethodManager).hideSoftInputFromWindow(this@SettingsActivity.etUnitInfo!!.windowToken, 0)
            }
            false
        }
        this.sbMapWidth!!.setOnSeekBarChangeListener(this)
        this.sbMapZoom!!.setOnSeekBarChangeListener(this)
        val metrics = DisplayMetrics()
        windowManager.defaultDisplay.getMetrics(metrics)
        val params = LayoutParams(-1, metrics.widthPixels)
        params.setMargins(10, 10, 10, 10)
        this.ivMap!!.layoutParams = params
        this.ivMap!!.viewTreeObserver.addOnGlobalLayoutListener(object : OnGlobalLayoutListener {
            override fun onGlobalLayout() {
                if (VERSION.SDK_INT < 16) {
                    this@SettingsActivity.ivMap!!.viewTreeObserver.removeGlobalOnLayoutListener(this)
                } else {
                    this@SettingsActivity.ivMap!!.viewTreeObserver.removeOnGlobalLayoutListener(this)
                }
                this@SettingsActivity.showMap()
            }
        })
    }

    override fun onResume() {
        super.onResume()
        this.mLocation!!.requestGPSUpdate()
    }

    override fun onStop() {
        super.onStop()
        this.mLocation!!.removeUpdates()
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.settings, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.save -> {
                val pref = MySharedPreferences.getInstance(this)
                pref.observerFilePath = this.etInputFolder!!.text.toString()
                pref.outputFilePath = this.etOutputFolder!!.text.toString()
                pref.unitInfo = this.etUnitInfo!!.text.toString()
                pref.mapWidth = this.sbMapWidth!!.progress + 100
                pref.mapZoom = this.sbMapZoom!!.progress
                pref.imageQuality = this.sbImageQuality!!.progress
                val returnIntent = Intent()
                returnIntent.putExtra(MainActivity.IS_UPDATE_SERVICE, true)
                setResult(-1, returnIntent)
                Toast.makeText(this, R.string.alert_save, Toast.LENGTH_LONG).show()
                finish()
                return true
            }
            else -> return super.onOptionsItemSelected(item)
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        if (ContextCompat.checkSelfPermission(this, "android.permission.ACCESS_FINE_LOCATION") == 0 || ContextCompat.checkSelfPermission(this, "android.permission.ACCESS_COARSE_LOCATION") == 0) {
            this.mLocation!!.removeUpdates()
        }
    }

    private fun showMap() {
        MyDownloader(this.mLocation!!.latitude, this.mLocation!!.longitude, this.sbMapZoom!!.progress, this.sbMapWidth!!.progress + 100, this.ivMap!!.height / 2, this.sbImageQuality!!.progress).execute(*arrayOfNulls(0))
    }

    override fun onProgressChanged(seekBar: SeekBar, progress: Int, fromUser: Boolean) {}

    override fun onStartTrackingTouch(seekBar: SeekBar) {}

    override fun onStopTrackingTouch(seekBar: SeekBar) {
        showMap()
    }

    override fun onSelectDirectory(path: String) {
        this.mDialog!!.dismiss()
        when (Integer.parseInt(this.mDialog!!.tag)) {
            R.id.etInputFolder /*2131492954*/ -> {
                this.etInputFolder!!.setText(path)
                return
            }
            R.id.etOutputFolder /*2131492956*/ -> {
                this.etOutputFolder!!.setText(path)
                return
            }
            else -> return
        }
    }

    override fun onCancelChooser() {
        this.mDialog!!.dismiss()
    }

    override fun onClick(v: View) {
        when (v.id) {
            R.id.etInputFolder /*2131492954*/ -> {
                showDialog(this.etInputFolder!!.text.toString(), R.id.etInputFolder.toString())
                return
            }
            R.id.etOutputFolder /*2131492956*/ -> {
                showDialog(this.etOutputFolder!!.text.toString(), R.id.etOutputFolder.toString())
                return
            }
            else -> return
        }
    }

    fun showDialog(folderPath: String, tag: String) {
        this.mDialog = DirectoryChooserFragment.newInstance(DirectoryChooserConfig.builder().newDirectoryName(getString(R.string.new_directory_name)).initialDirectory(folderPath).allowNewDirectoryNameModification(true).build())
        this.mDialog?.show(fragmentManager, tag)
    }
}
